akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2209,69 +2219,87 @@ private[akka] final class StatefulMap[S, In, Out](
     new GraphStageLogic(shape) with InHandler with OutHandler {
       lazy val decider: Decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
 
-      private var state: S = _
-      private var needInvokeOnCompleteCallback: Boolean = false
+      private var state: OptionVal[S] = _
 
       override def preStart(): Unit = {
-        state = create()
-        needInvokeOnCompleteCallback = true
+        state = OptionVal.Some(create())
+        throwIfNoState()
       }
 
-      override def onPush(): Unit =
+      override def onPush(): Unit = {
         try {
           val elem = grab(in)
-          val (newState, newElem) = f(state, elem)
-          state = newState
+          val (newState, newElem) = f(state.get, elem)
+          state = OptionVal.Some(newState)
+          throwIfNoState()
           push(out, newElem)
         } catch {
+          case ex: NullStateException => throw ex // don't cover with supervision
           case NonFatal(ex) =>
             decider(ex) match {
               case Supervision.Stop    => closeStateAndFail(ex)
               case Supervision.Resume  => pull(in)
-              case Supervision.Restart => resetStateAndPull()
+              case Supervision.Restart => restartState()
             }
         }
+      }
 
-      override def onUpstreamFinish(): Unit = closeStateAndComplete()
+      override def onUpstreamFinish(): Unit = {
+        completeStateIfNeeded() match {
+          case Some(elem) => emit(out, elem, () => completeStage())
+          case None       => completeStage()
+        }
+      }
 
       override def onUpstreamFailure(ex: Throwable): Unit = closeStateAndFail(ex)
 
       override def onDownstreamFinish(cause: Throwable): Unit = {
-        onComplete(state)
-        needInvokeOnCompleteCallback = false
+        completeStateIfNeeded()
         super.onDownstreamFinish(cause)
       }
 
-      private def resetStateAndPull(): Unit = {
-        needInvokeOnCompleteCallback = false
-        onComplete(state)
-        state = create()
-        needInvokeOnCompleteCallback = true;
-        pull(in)
-      }
-
-      private def closeStateAndComplete(): Unit = {
-        onComplete(state) match {
-          case Some(elem) => emit(out, elem, () => completeStage())
-          case None       => completeStage()
+      private def restartState(): Unit = {
+        completeStateIfNeeded() match {
+          case Some(elem) =>
+            emit(out, elem, { () =>
+              state = OptionVal.Some(create())
```

Comment:
In that PR we expect out to always be available, not sure why I thought it might, tracing it back it seems the only way here is if we didn't reach push in onPush so should always be ready to push, I'll change it to do direct push instead.